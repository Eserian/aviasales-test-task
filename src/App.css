.container {
  text-align: center;
}

.header {
  margin: 45px 0;
}

.grid-12 {
  display: grid;
  grid-template-columns: repeat(12, 1fr);
  grid-gap: 20px;
}

.col-4 {
  grid-column: span 4;
}

.col-8 {
  grid-column: span 8;
}

.filter {
  background: #fff;
  height: 250px;
  border: 1px solid #DFE5EC;
  box-shadow: 0px 2px 8px rgba(0, 0, 0, 0.1);
  border-radius: 5px;
}

.tabs {
  background: #fff;
  height: 50px;
  border: 1px solid #DFE5EC;
  margin-bottom: 20px;
  border-radius: 5px;
}

.ticket {
  height: 180px;
  background: #fff;
  box-shadow: 0px 2px 8px rgba(0, 0, 0, 0.1);
  border-radius: 5px;
}