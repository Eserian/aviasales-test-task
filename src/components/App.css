
.container {
  text-align: center;
}

.header {
  margin: 45px 0;
}

.main-grid {
  display: grid;
  grid-template-columns: repeat(12, 1fr);
  grid-gap: 20px;
}

.col-8 {
  grid-column: span 8;
}
