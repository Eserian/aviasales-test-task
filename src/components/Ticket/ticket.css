.ticket {
  display: flex;
  flex-direction: column;
  justify-content: center;
  height: 180px;
  background: #fff;
  box-shadow: 0px 2px 8px rgba(0, 0, 0, 0.1);
  border-radius: 5px;
  margin-top: 20px;
}

.ticket-header {
  display: flex;
}

.flight-data {
  display: flex;
  flex-direction: column;

}

.ticket-price {
  flex: 1 1 auto;
  text-align: left;
  font-style: normal;
  font-weight: 600;
  font-size: 24px;
  line-height: 24px;
  color: #2196F3;
  padding-left: 20px;
}

.ticket-logo {
  flex: 1 1 auto;
  text-align: right;
  padding-right: 50px;
}

.title {
  font-style: normal;
  font-weight: 600;
  font-size: 12px;
  line-height: 18px;
  letter-spacing: 0.5px;
  text-transform: uppercase;
  color: #A0B0B9;
  text-align: left;
}

.row {
  display: flex;
}

.subtitle {
  font-style: normal;
  font-weight: 600;
  font-size: 14px;
  line-height: 21px;
  color: #4A4A4A;
  text-align: left;
}

.container {
  margin-top: 10px;
  flex: 1 1 33.333%;
  padding-left: 20px;
}
