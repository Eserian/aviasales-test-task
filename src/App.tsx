import React from 'react';
import logo from './logo.svg';
import './App.css';

function App() {
  return (
    <div className="container">
      <header className="header">
        <img src={logo} className="App-logo" alt="logo" />
      </header>
      <main className="grid-12">
        <div className="filter col-4"></div>
        <div className="col-8">
          <div className="tabs"></div>
          <div className="tickets">
            <div className="ticket"></div>
          </div>
        </div>
      </main>
    </div>
  );
}

export default App;
