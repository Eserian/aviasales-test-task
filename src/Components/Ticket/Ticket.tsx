import React, { FC }from 'react';
import { ticket } from '../App';
import moment from 'moment';
import './ticket.css';

type FlightCardProps ={
  ticket: ticket
};

const stopsConvertToText = (stops: string[]) => {
  switch (stops.length) {
    case 0:
      return 'Без пересадок';
    case 1:
      return '1 пересадка';
    default:
      return `${stops.length} пересадки`
  } 
};

const getTimeFromMins = (mins: number) => {
  const hours: number = Math.trunc(mins/60);
  const minutes: number = mins % 60;
  return `${hours}ч ${minutes}м`;
};

export const Ticket: FC<FlightCardProps> = ({ ticket }) => {
  return (
    <div className="ticket">
      <header className="ticket-header">
        <div className="ticket-price">{`${ticket.price} Р`}</div>
        <div className="ticket-logo"><img src={`//pics.avs.io/99/36/${ticket.carrier}.png`} alt="logo"/></div>
      </header>
      <div className="flight-data">
        <div className="row">
          <div className="container">
            <div className="title">{`${ticket.segments[0].origin}-${ticket.segments[0].destination}`}</div>
            <div className="subtitle">{`${moment.utc(ticket.segments[0].date).format('HH:mm')} - ${moment.utc(ticket.segments[0].date).add(ticket.segments[0].duration, 'm').format('HH:mm')}`}</div>
          </div>
          <div className="container">
            <div className="title">В пути</div>
            <div className="subtitle">{getTimeFromMins(ticket.segments[0].duration)}</div>
          </div>
          <div className="container">
            <div className="title">{stopsConvertToText(ticket.segments[0].stops)}</div>
            <div className="subtitle">{ticket.segments[0].stops.join(',')}</div>
          </div>
        </div>
        <div className="row">
          <div className="container">
          <div className="title">{`${ticket.segments[1].origin}-${ticket.segments[1].destination}`}</div>
            <div className="subtitle">{`${moment.utc(ticket.segments[1].date).format('HH:mm')} - ${moment.utc(ticket.segments[1].date).add(ticket.segments[1].duration, 'm').format('HH:mm')}`}</div>
          </div>
          <div className="container">
            <div className="title">В пути</div>
            <div className="subtitle">{getTimeFromMins(ticket.segments[1].duration)}</div>
          </div>
          <div className="container">
            <div className="title">{stopsConvertToText(ticket.segments[1].stops)}</div>
            <div className="subtitle">{ticket.segments[1].stops.join(',')}</div>
          </div>
        </div>
      </div>
    </div>
  )
}