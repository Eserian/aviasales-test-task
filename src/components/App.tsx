import React, { FC, useEffect, useState } from 'react';
import './App.css';
import { Header } from './Header/Header';
import { Filter } from './Filter/Filter';
import { Sorting } from './Sorting/Sorting';
import { Ticket } from './Ticket/Ticket';
import { Preloader } from './Preloader/Preloader';
import axios from 'axios';

type flight = {
  origin: string
  destination: string
  date: string
  stops: string[]
  duration: number
}

export type ticket = {
  price: number
  carrier: string
  segments: flight[]
}

type sortingMap = {
  [key: string]: (a: ticket, b: ticket) => number
}

const comparePrice = (a: ticket, b: ticket) => a.price - b.price;

const compareFlightDuration = (a: ticket, b: ticket) => {
  const firstFlightDuration = a.segments[0].duration + a.segments[1].duration;
  const secondFlightDuration = b.segments[0].duration + b.segments[1].duration;

  return firstFlightDuration - secondFlightDuration;
}

const sortingMap: sortingMap = {
  cheap: comparePrice,
  fast: compareFlightDuration
}

const API_URL = 'https://front-test.beta.aviasales.ru';

const getSearchId = async () => {
  const url = `${API_URL}/search`;
  const response = await axios.get(url);
  return response.data.searchId;
}

const getTicketPack = async (searchId: string) => {
  const url = `${API_URL}/tickets?searchId=${searchId}`;
  const response = await axios.get(url);
  return response.data;
}

const App: FC = () => {

  const [allTickets, setAllTickets] = useState([] as ticket[]);
  const [isLoading, setIsLoading] = useState(true);
  const [sort, setSort] = useState('cheap');
  const [filterParams, setFilterParams] = useState([0, 1, 2, 3]);

  const filter = (ticket: ticket) => {
    const flightForceStops = ticket.segments[0].stops.length;
    const flightBackStops = ticket.segments[1].stops.length;
    return filterParams.includes(flightForceStops) && filterParams.includes(flightBackStops);
  }

  type iter = (searchId: string, acc: ticket[]) => Promise<ticket[] | iter>

  useEffect(() => {
    const loadTickets = async () => {
      const iter: iter = async (searchId: string, acc: ticket[]) => {
        try {
          const ticketPack = await getTicketPack(searchId);
          const newAcc = [...acc, ...ticketPack.tickets];
          if (ticketPack.stop) {
            return newAcc;
          }
          return iter(searchId, newAcc);
        } catch (e) {
          return iter(searchId, acc);
        }
      }

      const searchId = await getSearchId();
      const tickets = await iter(searchId, []);

      setAllTickets(tickets as ticket[]);
      setIsLoading(false);
    }
    loadTickets();
  }, []);

  return (
    <>
      <Header />
      {
        isLoading ?
          <Preloader /> :
          <main className="main-grid">
            <Filter handleFilter={setFilterParams} />
            <div className="col-8">
              <Sorting handleSort={setSort} />
              <div className="ticketList">
                {
                  allTickets
                    .filter(filter)
                    .sort(sortingMap[sort])
                    .slice(0, 5)
                    .map((ticket: ticket, i) => <Ticket key={i} data={ticket} />)
                }
              </div> 
            </div>
          </main>
      }
    </>
  );
}

export default App;
