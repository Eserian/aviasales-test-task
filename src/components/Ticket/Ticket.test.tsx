import React from 'react';
import { render } from '@testing-library/react';
import { Ticket } from './Ticket';
import { ticket } from '../App';

const testTicket: ticket = {
  price: 93332,
  carrier: "EY",
  segments: [
    {
      origin: "MOW",
      destination: "HKT",
      date: "2020-09-21T05:35:00.000Z",
      stops: ["IST","HKG"],
      duration: 1706
    },
    {
      origin: "HKT",
      destination: "MOW",
      date: "2020-10-11T14:56:00.000Z",
      stops: ["DXB"],
      duration: 681
    }
  ]
}

test('Render correctly', () => {
  const ticket = render(<Ticket data={testTicket} />);
  expect(ticket).toMatchSnapshot();
})
