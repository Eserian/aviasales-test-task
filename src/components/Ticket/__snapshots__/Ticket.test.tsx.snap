// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="ticket"
      >
        <header
          class="ticket-header"
        >
          <div
            class="ticket-price"
          >
            93332 Р
          </div>
          <div
            class="ticket-logo"
          >
            <img
              alt="logo"
              src="//pics.avs.io/99/36/EY.png"
            />
          </div>
        </header>
        <div
          class="flight-data"
        >
          <div
            class="row"
          >
            <div
              class="container"
            >
              <div
                class="title"
              >
                MOW-HKT
              </div>
              <div
                class="subtitle"
              >
                05:35 - 10:01
              </div>
            </div>
            <div
              class="container"
            >
              <div
                class="title"
              >
                В пути
              </div>
              <div
                class="subtitle"
              >
                28ч 26м
              </div>
            </div>
            <div
              class="container"
            >
              <div
                class="title"
              >
                2 пересадки
              </div>
              <div
                class="subtitle"
              >
                IST,HKG
              </div>
            </div>
          </div>
          <div
            class="row"
          >
            <div
              class="container"
            >
              <div
                class="title"
              >
                HKT-MOW
              </div>
              <div
                class="subtitle"
              >
                14:56 - 02:17
              </div>
            </div>
            <div
              class="container"
            >
              <div
                class="title"
              >
                В пути
              </div>
              <div
                class="subtitle"
              >
                11ч 21м
              </div>
            </div>
            <div
              class="container"
            >
              <div
                class="title"
              >
                1 пересадка
              </div>
              <div
                class="subtitle"
              >
                DXB
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="ticket"
    >
      <header
        class="ticket-header"
      >
        <div
          class="ticket-price"
        >
          93332 Р
        </div>
        <div
          class="ticket-logo"
        >
          <img
            alt="logo"
            src="//pics.avs.io/99/36/EY.png"
          />
        </div>
      </header>
      <div
        class="flight-data"
      >
        <div
          class="row"
        >
          <div
            class="container"
          >
            <div
              class="title"
            >
              MOW-HKT
            </div>
            <div
              class="subtitle"
            >
              05:35 - 10:01
            </div>
          </div>
          <div
            class="container"
          >
            <div
              class="title"
            >
              В пути
            </div>
            <div
              class="subtitle"
            >
              28ч 26м
            </div>
          </div>
          <div
            class="container"
          >
            <div
              class="title"
            >
              2 пересадки
            </div>
            <div
              class="subtitle"
            >
              IST,HKG
            </div>
          </div>
        </div>
        <div
          class="row"
        >
          <div
            class="container"
          >
            <div
              class="title"
            >
              HKT-MOW
            </div>
            <div
              class="subtitle"
            >
              14:56 - 02:17
            </div>
          </div>
          <div
            class="container"
          >
            <div
              class="title"
            >
              В пути
            </div>
            <div
              class="subtitle"
            >
              11ч 21м
            </div>
          </div>
          <div
            class="container"
          >
            <div
              class="title"
            >
              1 пересадка
            </div>
            <div
              class="subtitle"
            >
              DXB
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
